builtin/mainmenu/misc.lua

```diff
@@ -4,3 +4,21 @@
 function core.sound_stop(handle, ...)
 	return handle:stop(...)
 end
+
+-- On Android, closing the app via the "Recents" screen won't result in a clean
+-- shutdown, discarding any setting changes made by the user.
+-- To avoid that, we write the settings file in more cases on Android.
+-- We don't do this on desktop platforms because you can have multiple
+-- instances of Minetest running at the same time there.
+function write_settings_if_android()
+	if PLATFORM == "Android" then
+		core.settings:write()
+	end
+end
+
+local old_core_start = core.start
+core.start = function()
+	-- Necessary for saving the selected world, server address, player name, etc.
+	write_settings_if_android()
```

Comment:
could be moved here: https://github.com/minetest/minetest/blob/master/src/client/clientlauncher.cpp#L558

and IMO it wouldn't be bad to save it on all platforms

Reply:
both suggestion are no problem to implement, but why?

Reply:
1: I think it's cleaner than wrapping the function in Lua

2: When you join a game you switch to another part of the engine. I think that's a good point to save the settings anyway.
If you have multiple clients open they will overwrite each others files anyway so that doesn't matter.

> in all cases covered by this PR or only in this specific case?

only this one

Reply:
I've tried making some changes to `clientlauncher.cpp`, and it's horrible. What's wrong with that file?

This saves the settings twice on shutdown, but as far I've tested, the condition is never triggered anyway:

https://github.com/minetest/minetest/blob/a8cf10b0b523bf4c3234be29b0dd51e719cb00c1/src/client/clientlauncher.cpp#L252-L257

`launch_game` doesn't actually launch a game. It launches the mainmenu and returns a boolean indicating whether we're supposed to launch a game.

`game_has_run` doesn't store whether the game has run, but whether it will run (whether we're supposed to start a game):

https://github.com/minetest/minetest/blob/a8cf10b0b523bf4c3234be29b0dd51e719cb00c1/src/client/clientlauncher.cpp#L233-L234

This sets a setting to which there is no second reference anywhere in the codebase (it was supposedly used by the C++ mainmenu):

https://github.com/minetest/minetest/blob/a8cf10b0b523bf4c3234be29b0dd51e719cb00c1/src/client/clientlauncher.cpp#L443-L444

This ...

1. ... checks whether the gamespec is invalid, and aborts launching a game in that case.
2. ... checks whether `porting::signal_handler_killstatus()` returns a non-null pointer and decides to launch a game in that case. (`porting::signal_handler_killstatus()` always returns a non-null pointer.)
3. ... checks again whether the gamespec is invalid, and aborts launching a game in that case. (Checking again is 100% redundant.) (This never runs anyway because we always already return at step 2.)

https://github.com/minetest/minetest/blob/a8cf10b0b523bf4c3234be29b0dd51e719cb00c1/src/client/clientlauncher.cpp#L507-L524

In the same function a bit above however, the `porting::signal_handler_killstatus()` check is done correctly:

https://github.com/minetest/minetest/blob/a8cf10b0b523bf4c3234be29b0dd51e719cb00c1/src/client/clientlauncher.cpp#L422-L424

Reply:
Okay, updated (see the diff). What do you think?